src/network/network_coordinator.cpp

```diff
@@ -263,6 +304,49 @@ bool ClientNetworkCoordinatorSocketHandler::Receive_GC_DIRECT_CONNECT(Packet *p)
 	return true;
 }
 
+bool ClientNetworkCoordinatorSocketHandler::Receive_GC_STUN_REQUEST(Packet *p)
+{
+	std::string token = p->Recv_string(NETWORK_TOKEN_LENGTH);
+
+	this->stun_handlers[token][AF_INET6] = ClientNetworkStunSocketHandler::Stun(token, AF_INET6);
+	this->stun_handlers[token][AF_INET] = ClientNetworkStunSocketHandler::Stun(token, AF_INET);
+	return true;
+}
+
+bool ClientNetworkCoordinatorSocketHandler::Receive_GC_STUN_CONNECT(Packet *p)
+{
+	std::string token = p->Recv_string(NETWORK_TOKEN_LENGTH);
+	uint8 tracking_number = p->Recv_uint8();
+	uint8 family = p->Recv_uint8();
+	std::string host = p->Recv_string(NETWORK_HOSTNAME_PORT_LENGTH);
```

Comment:
https://en.wikipedia.org/wiki/STUN mentions that the IP address and port should be obfuscated (in their example by a XOR) so deep inspection stuff does (likely) not mess with the data. Might that be a wise strategy to employ here? Maybe not XOR but some other method? Given everything's a character, maybe add 10 to every character or so?

Reply:
lol ... I wouldn't even know why a NAT would waste CPU cycles on that, but sure.

Most changes are a bit tricky, as it needs to remain a string .. so I will just reverse the hostname. That should be sufficient I hope :D

Reply:
ah, no, this is not needed at all. Pfew, saves me from weird code.

Normally what STUN does, is return to the client itself what his public IP:port is for that socket. For OpenTTD I opt'd not to do that, as that would mean we need another round-trip from the client to the Game Coordinator to tell about that public IP:port. I found that wasteful and potentially allows OpenTTD to be used for DoS attacks. So, instead, the STUN server tells the GC about the public IP:port, and this is never told back to the client.

The client does get back the IP:port of its peer, but no NAT will rewrite that. So there is no need to mitigate this in our setup.